spec/std/set_spec.cr

```diff
@@ -179,7 +179,7 @@ describe "Set" do
   end
 
   it "does -" do
-    set1 = Set{1, 2, 3, 4, 5}
+    set1 = Set(Int32 | Char){1, 2, 3, 4, 5}
```

Comment:
It bugs me how strict you need to be in this examples. I imagine that `set1 - set2` no longer compiles when `set1 : Set(Int32)` and `set2 : Enumerable(Int32 | Char)`. To me it falls into the same behaviour we want on equality: although we can infer it will always be false due to typing missmatch, we want to make it a valid that will return false instead of not compile.

It could happen that this set for example is generated from templates, macros and the explicit typing will be hard to inject. Hence, the type will be the infered from the elements, leading to `Set(Int32)`.

Reply:
I agree, this is a a pretty irritating tradeoff on paper, but let’s think about the overall impact of this consequence. How often is it that you have a homogeneous set that you subtract a heterogeneous set from?

Honestly, I can’t think of a single time I’ve ever used a heterogeneous set in application code, let alone in this particular scenario where I’m subtracting one from a homogenous set. To be clear, you don’t need to specify types for _all_ sets. Just ones where the type inference doesn’t get you what you need.

I do think that if you’re going to compare values, it’s probably a good idea to know that the types you’re comparing against can be different and the type specifier helps ensure that not only do you recognize that now, but you’ll recognize that when you change the code later. My perspective is that it’s not the same as adding two sets of different types where the result is a union of all the types in both sets. If there’s a bug there it’ll likely be caught elsewhere in the app because of the new type of the resulting set. For example:

```crystal
numbers = Set{1, 2, 3}
another_set = Set{“1”, “2”}

result = numbers - another_set # => Set(Int32){1, 2, 3}
result = numbers + another_set # => Set(Int32 | String){1, 2, 3, “1”, “2”}
```

I don’t think the `-` version makes any sense at all and if I’m doing that, it’s probably a bug. When adding them, it changes the type and is more likely to catch that bug. With this PR, the subtraction would be a bug.

IMO, the type checker is there to help you. It’s why a lot of people still use static languages. Something that becomes a compile-time error that could be a runtime error is often a good thing. Being able to tell the compiler you know what you’re doing by simply adding a type annotation is also good.

All that said, I found a way to let me remove the explicit type specifier. I don’t think I agree with it, but it keeps the same behavior in this scenario.

Reply:
So I guess we should disallow `-` between sets of different types? That is, you can only do A - B if A's element types are a superset of B.

That would need to apply to all collection types.

The problem is always equality and inclusion, specially regarding `==`. Because `1` == `1.0`, and because `==` can be generally defined between any two types, when you have `Set{1, 2, 3}` then subtracting `Set{1.0}` could be valid...

Interestingly, in Ruby:

```
Set[1, 2, 3] - Set[1.0, 2.0, 3.0] # => #<Set: {1, 2, 3}>
```

Also:

```
irb(main):004:0> h = {1 => 2}
=> {1=>2}
irb(main):005:0> h[1.0]
=> nil
```

That's because Hash uses `eql?` and not `==` for comparing key values.

But then also in Ruby:

```
{a: 1} == {a: 1.0} # => true
```

but that's because it uses `==` for comparing elements, and `1 == 1.0 # => true`.

So maybe we can disallow `h[1.0]` for `Hash(Int32, V)`. However, if you have `Hash(Int32 | Float64)` and you have `1` as a key and you ask for `1.0` it will work and return the value even though with `Hash(Int32, V)` it wouldn't work... because we don't make the distinction of `==` and `eql?` like Ruby.

This is a hard problem... :-)

We'll need to decide if we want to introduce `eql?` to types, which might make sense in the context of union types and the way Ruby works.

Maybe we can also disallow `==` between any two types. For example comparing `Int32` vs `String` wouldn't compile, which might make sense. But we can allow comparison between any two reference types, that being `false` by default.

Sorry for the wall of text... my point is that this PR can't be thought in isolation will all of these other things. And changing that is a huge breaking change which I'm not sure we can/should do right before 1.0.

Reply:
> So I guess we should disallow - between sets of different types? That is, you can only do A - B if A's element types are a superset of B.

In an ideal world, I think that makes sense.

> That would need to apply to all collection types.

Yeah, that's the catch. 🙂 It's why, even though I disagree with @bcardiff I still made it work without the explicit type annotation.

Your focus on consistency is important and I'm glad you called it out. Consistency can often be more important than anyone's perspective of what's "correct" because it allows you to guess at how something works. That's a super power for a programming language. Having to consult documentation all the time gets old quick.

I definitely want to have the conversation about whether `Array(Int32)#-(Array(String))` should compile, but that particular change is too far outside the scope of this PR and I apologize for hammering on it. All I wanted was for the compiler to tell me when I changed my hash key type that I'd missed some of my key checks. :joy:

>  if you have Hash(Int32 | Float64) and you have 1 as a key and you ask for 1.0 it will work and return the value even though with Hash(Int32, V) it wouldn't work... because we don't make the distinction of == and eql? like Ruby.

I've been using Ruby for 16 years and, honestly, I never understood why Ruby used `eql?` for that. But this _definitely_ sounds like it might be why.

Reply:
I'm loving the discussion here. I also use Crystal in large part due to the type-safety and would prefer the A must be a superset of B. I've been bitten by things like this when developing the internals of Lucky where I'm working with data from outside the system (cookies, params, sql).

I personally would love a more strict equality with an "escape hatch" of some kind. Maybe that's with `eql?` or a different method. The key is that you have that escape hatch. So strict by default to catch bugs, with an escape hatch for if you want to get into the kinky stuff.

You could even do something similar to what we do in Lucky where there is a fallback method that generates a nice little error telling you why it happened and what you can do isntead:

```
Can't compare/set ekys/etc Objects of different types.

Make sure the objects are the type you expect, or use `some_other_method` to allow setting/getting/comparing objects of different types.

This way you get the safety, but you have the escape hatch along with a nice error. I think this will really make the lagnuage super powerful and easy for beginners.
```

Reply:
This would require something like https://github.com/crystal-lang/crystal/issues/6997#issuecomment-450353013:

```crystal
struct Set(T)
  def -(other : Set(U)) forall U <= T
    # ...
  end

  @[Deprecated("Ensure *other*'s element type is `T` or a subtype of it")]
  def -(other : Set)
    dup
  end
end
```

The story is more complex if we want to allow unions of `Set` instances as arguments (mainly due to #8973).